sdk/storage_datalake/src/shared_key_authorization_policy.rs

```diff
@@ -0,0 +1,244 @@
+        .unwrap_or("");
+    format!(
+        "{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}\n{}{}",
+        http_method.as_str(),
+        add_if_exists(http_headers, http::header::CONTENT_ENCODING),
+        add_if_exists(http_headers, http::header::CONTENT_LANGUAGE),
+        cl,
+        add_if_exists(http_headers, azure_storage::headers::CONTENT_MD5),
+        add_if_exists(http_headers, http::header::CONTENT_TYPE),
+        add_if_exists(http_headers, http::header::DATE),
+        add_if_exists(http_headers, http::header::IF_MODIFIED_SINCE),
+        add_if_exists(http_headers, http::header::IF_MATCH),
+        add_if_exists(http_headers, http::header::IF_NONE_MATCH),
+        add_if_exists(http_headers, http::header::IF_UNMODIFIED_SINCE),
+        add_if_exists(http_headers, http::header::RANGE),
+        canonicalize_header(http_headers),
+        canonicalized_resource(storage_account_name, url)
+    )
+
+    // expected
+    // GET\n /*HTTP Verb*/
+    // \n    /*Content-Encoding*/
+    // \n    /*Content-Language*/
+    // \n    /*Content-Length (include value when zero)*/
+    // \n    /*Content-MD5*/
+    // \n    /*Content-Type*/
+    // \n    /*Date*/
+    // \n    /*If-Modified-Since */
+    // \n    /*If-Match*/
+    // \n    /*If-None-Match*/
+    // \n    /*If-Unmodified-Since*/
+    // \n    /*Range*/
+    // x-ms-date:Sun, 11 Oct 2009 21:49:13 GMT\nx-ms-version:2009-09-19\n
+    //                                  /*CanonicalizedHeaders*/
+    // /myaccount /mycontainer\ncomp:metadata\nrestype:container\ntimeout:20
+    //                                  /*CanonicalizedResource*/
+    //
+    //
+}
+
+fn add_if_exists<K: http::header::AsHeaderName>(h: &HeaderMap, key: K) -> &str {
+    match h.get(key) {
+        Some(ce) => ce.to_str().unwrap(),
+        None => "",
+    }
+}
+
+fn canonicalize_header(h: &HeaderMap) -> String {
+    let mut v_headers = h
+        .iter()
+        .filter(|(k, _v)| k.as_str().starts_with("x-ms"))
+        .map(|(k, _)| k.as_str())
+        .collect::<Vec<_>>();
+    v_headers.sort_unstable();
+
+    let mut can = String::new();
+
+    for header_name in v_headers {
+        let s = h.get(header_name).unwrap().to_str().unwrap();
+        can = can + header_name + ":" + s + "\n";
+    }
+    can
+}
+
+fn canonicalized_resource(account: &str, u: &url::Url) -> String {
+    let mut can_res: String = String::new();
+    can_res += "/";
+    can_res += account;
+
+    let paths = u.path_segments().unwrap();
+
+    for p in paths {
+        can_res.push('/');
+        can_res.push_str(&*p);
+    }
+    can_res += "\n";
+
+    // query parameters
+    let query_pairs = u.query_pairs(); //.into_owned();
+    {
+        let mut qps = Vec::new();
+        {
+            for (q, _p) in query_pairs {
+                trace!("adding to qps {:?}", q);
+
+                // add only once
+                if !(qps.iter().any(|x: &String| x == q.as_ref())) {
```

Comment:
Use a `HashSet` instead of a `Vec` and you uniqueness is guaranteed.

Reply:
For this PR, I'd like to change as little as possible in this code, since it's directly copied from the pre-pipeline shared key auth code in storage_account_client.rs. If there are any issues there, I'd like to address them in a follow-up PR.